Foundation/Stream.swift

```diff
@@ -225,6 +228,43 @@ open class OutputStream : Stream {
     }
 }
 
+extension InputStream {
+    func seek(to position: UInt64) throws {
+        guard position > 0 else {
+            return
+        }
+        
+        guard position < Int.max else { throw _Error.cantSeekInputStream }
+        
+        let bufferSize = 1024
+        var remainingBytes = Int(position)
+        
+        let buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: bufferSize, alignment: MemoryLayout<UInt8>.alignment)
+        
+        guard let pointer = buffer.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
+            buffer.deallocate()
+            throw _Error.cantSeekInputStream
+        }
+        
+        if self.streamStatus == .notOpen {
+            self.open()
+        }
+        
+        while remainingBytes > 0 && self.hasBytesAvailable {
```

Comment:
So for the seek to work, all the bytes in the seek have to be buffered? Is that the contract? If the bytes aren't available (yet) then this code will throw below on if remainingBytes != 0. Is that the intent? Does seek need to accommodate the case where some of the range has not yet been received over the network?

Reply:
> So for the seek to work, all the bytes in the seek have to be buffered? Is that the contract? 

Only first n bytes, and size of buffer is 1024bytes not n bytes. We just use it as temporary storage for `InputStream::read`. I believe this is easy way to seek InputStream. 

>  If the bytes aren't available (yet) then this code will throw below on if remainingBytes != 0. Is that the intent? 

Yep, we throws error as you mention, and(for current pr) return `CURL_SEEKFUNC_CANTSEEK` to curl which tell libcurl seeking can't be done, so libcurl might try other means instead.

> Does seek need to accommodate the case where some of the range has not yet been received over the network?

I hope yes, but not for this extension. We just rely on `read` function and do no dive into implementation of read function. If no bytes available or some ranges of bytes not being receiving yet, we just throws `cantSeekInputStream`. In scope of this PR we just just need to return to curl `CURL_SEEKFUNC_CANTSEEK`. 
 Also I want to add:

This a simple seek implementation which not handles cases if some bytes have not being received yet. This seek function currently just used in curl callback. It seems that I need to rename function to `seekIfBytesAvailable`, or move it to `NativeProtocol:: seekInputStream`(but it will be hard to test).

Reply:
It seems that I need to move away `seek` function from InputStream 😅

Reply:
@jdberry can you please describe how do we can resolve this? thanks!

Reply:
Hi @albertaleksieiev. I won't profess to fully understand the code or the requirements of the code or the exactly what the contract of this seek functionality is. So it's possible I'm totally off base.

My guess would be that since you're seeking forward in the stream, and since not all the bytes to seek past may yet have arrived, you'd have to set an instance variable that would track bytes to be dumped or ignored as they arrive in the future. As future bytes arrive, they would be consumed (ignored) until the seek is fulfilled. There might be other operations (future seeks, others?) that would reset the seek count before it is fulfilled. 

I'll reinforce that this matches my naive expectation for what it sounds like this functionality would do. I am not however an expert on this code and I may be completely misinterpreting (and overcomplicating?) the expected outcome.

Reply:
And, @albertaleksieiev, above you wrote:

>  In scope of this PR we just just need to return to curl `CURL_SEEKFUNC_CANTSEEK`.

If you really feel that works, that may be all that's needed. I was really just wanting assurance that you or somebody else has looked into it and this implementation is sufficient to meet the requirements.